src/stream/src/executor/dml.rs

```diff
@@ -151,8 +182,41 @@ impl DmlExecutor {
                         TxnMsg::End(txn_id) => {
                             let mut txn_buffer = active_txn_map.remove(&txn_id)
                                 .unwrap_or_else(|| panic!("Receive an unexpected transaction end message. Active transaction map doesn't contain this transaction txn_id = {}.", txn_id));
-                            for chunk in txn_buffer.vec.drain(..) {
-                                yield Message::Chunk(chunk);
+
+                            let txn_buffer_cardinality = txn_buffer
+                                .vec
+                                .iter()
+                                .map(|c| c.cardinality())
+                                .sum::<usize>();
+                            let batch_group_cardinality =
+                                batch_group.iter().map(|c| c.cardinality()).sum::<usize>();
+
+                            if txn_buffer_cardinality >= self.chunk_size {
+                                // txn buffer isn't small, so yield.
+                                for chunk in txn_buffer.vec {
+                                    yield Message::Chunk(chunk);
+                                }
+                            } else if txn_buffer_cardinality + batch_group_cardinality
+                                <= self.chunk_size
+                            {
+                                // txn buffer is small and batch group has space.
+                                batch_group.extend(txn_buffer.vec);
+                            } else {
+                                // txn buffer is small and batch group has no space, so yield the large one.
+                                if txn_buffer_cardinality < batch_group_cardinality {
+                                    mem::swap(&mut txn_buffer.vec, &mut batch_group);
+                                }
```

Comment:
IIUC the inserted records might be reordered here, which seems to be inconsistent with this discussion? https://github.com/risingwavelabs/risingwave/pull/13872#discussion_r1423853229

Reply:
Current implementation at least guarantees that data from the same transaction would be sent to downstream at the same time. Otherwise, part of data from the same transaction would be sent to downstream at first and the other part is kept in the `batch_group`.

Reply:
Ok, sounds acceptable to me